index.d.ts

```diff
@@ -0,0 +1,41 @@
+import * as React from 'react';
+import Store from '@orbit/store';
+import { TransformOrOperations, QueryBuilder, Source } from '@orbit/data';
+
+
+export interface DataProviderProps {
+  dataStore: Store;
+}
+
+export interface RecordsToProps {
+  [key: string]: (q: QueryBuilder) => any;
+}
+
+export interface WithData {
+  dataStore: Store;
+  sources: { [sourceName: string]: Source };
+}
+
+export type WithDataProps =
+  & {
+    queryStore: (transformOrOperations: TransformOrOperations, options?: object, id?: string) => any;
+    updateStore: (transformOrOperations: TransformOrOperations, options?: object, id?: string) => any;
+    dataStore: Store;
+    sources: { [sourceName: string]: Source };
+  }
+  & WithData
+
+
+type MapRecordsToPropsFn = (...args: any[]) => RecordsToProps;
+
+export type MapRecordsToProps =
+  | RecordsToProps
+  | MapRecordsToPropsFn
+
+export class DataProvider extends React.Component<DataProviderProps> {}
+
+
+export function withData<TWrappedProps>(mapRecordsToProps: MapRecordsToProps):
+  <Props, State>(
+    WrappedComponent: React.Component<any, any, any> & { setState(): void}
```

Comment:
Just out of curiosity: why is the setState() definition here?

Reply:
The most annoying thing about the react + typescript combo is how react needs to identify classes / components. I actually super hate this behavior.

Basically, sometimes there is a requirement on a component to be stateful, when it doesn't matter for your interaction with that component. I feel like the interfaces implemented in the type reacts defs are maybe incorrect